File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0, right = 0, maxLen = 0;
        int count = s.length();
        int[] map = new int[128]; //ASCII
        if(s.length() <= 0){
            return 0;
        }

        while (right < s.length()){
            char rightChar = s.charAt(right);
            map[rightChar]++;

            while(map[rightChar] > 1){ //duplcate char
                char leftChar = s.charAt(left);
                map[leftChar]--; //shrink window from left to potentially remove the duplicate
                left++;
            }
            maxLen = Math.max(maxLen, right - left + 1);
            right++;
        }
        return maxLen;   
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        var left = 0
        var right = 0
        var maxLen = 0
        var count = s.count
        var map = [Int](repeating: 0, count: 128)
        let chars = Array(s)

        if s.count <= 0{return 0}

        while right<s.count{
            var rightChar = chars[right]
            map[Int(rightChar.asciiValue!)] += 1

            while map[Int(rightChar.asciiValue!)] > 1{
                var leftChar = chars[left]
                map[Int(leftChar.asciiValue!)] -= 1
                left += 1
            }
            maxLen =  max(maxLen, right - left + 1)  
            right += 1     
        }
        return maxLen
    }
}